Compound sprite nesting

i25db opened the pull request:
This PR implements the ability to nest CompoundSprites. It also adds as comprehensive of a test suite as I could come up with to cover the different combinations of panning and zooming of nested sprites.

I did not adhere to any particular naming convention or style guidelines so if that is desired just let me know and I can make adjustments.

Commit:
First attempt with dest_rect

Commit:
First attempt with dest_rect

Commit:
Extract CompoundSprite transformation into its own method

Commit:
Add initial test for CompoundSprite

Commit:
Add initial test for CompoundSprite

Commit:
Add full test suite for CompoundSprite#transform

Commit:
Use CompoundSprite#transform in CompoundSprite#draw_override

danhealy commented on lib/zif/compound_sprite.rb:
```diff
@@ -103,14 +103,13 @@ def source_rect
     # @api private
-    def draw_override(ffi_draw)
+    def draw_override(ffi_draw, parent_transform={offset: {x: 0, y: 0}, zoom: {x: 1.0, y: 1.0}})
       # $services&.named(:tracer)&.mark("CompoundSprite(#{@name})#draw_override: begin")
       # Treat an alpha setting of 0 as an indication that it should be hidden, to match Sprite behavior
       return if @a.zero?
 
-      view_actual_size! unless source_is_set?
+      trans = transform(parent_transform)
```
I am a little concerned about the allocation of objects here, because this code path can be the performance bottleneck for games using large numbers of sprites in activelayers.  Since this is a private API I'd like to use a little more complex pattern of individual params instead of allocating hashes. `trans_offset_x: 0, trans_offset_y: 0, trans_zoom_x: 1.0, trans_zoom_y: 1.0`

Then we should calculate the transformation of each individually; instead of in a single method returning an object we might need 4 methods returning floats.

This removes the need to perform `{}.merge()` in the calculation as well.

Commit:
Refactor #transform and #draw_override

CompoundSprite#transform and #draw_override now
accept individual parameters instead of hashes

i25db committed:
Merge pull request #1 from danhealy/main

Merge changes from danhealy/dragonruby-zif
